orderer/common/follower/block_puller.go

```diff
@@ -24,73 +24,152 @@ import (
 type ChannelPuller interface {
 	PullBlock(seq uint64) *common.Block
 	HeightsByEndpoints() (map[string]uint64, error)
+	UpdateEndpoints(endpoints []cluster.EndpointCriteria)
 	Close()
 }
 
-// BlockPullerFromJoinBlock creates a new block puller from a join-block.
-// This block puller is used for on-boarding, when join-block.Number >= ledger.Height().
-// This block puller can fetch the genesis block and skip verifying it.
-func BlockPullerFromJoinBlock(
-	joinBlock *common.Block,
+//go:generate counterfeiter -o mocks/block_puller_factory.go -fake-name BlockPullerFactory . BlockPullerFactory
+
+type BlockPullerFactory interface {
+	BlockPuller(configBlock *common.Block) (ChannelPuller, error)
+	UpdateVerifierFromConfigBlock(configBlock *common.Block) error
+}
+
+// BlockPullerCreator creates a BlockPuller on demand.
+// It also maintains a link to a block signature verifier, and exposes a method to update it on incoming config blocks.
+// The BlockPuller generated by this factory always accesses the updated verifier, since it is generated
+// with a link to the factory's VerifyBlockSequence method.
+type BlockPullerCreator struct {
+	channelID               string
+	bccsp                   bccsp.BCCSP
+	blockSigVerifierFactory cluster.VerifierFactory // Creates a new block signature verifier
+	blockSigVerifier        cluster.BlockVerifier   // The current block signature verifier, from the latest channel config
+	clusterConfig           localconfig.Cluster
+	signer                  identity.SignerSerializer
+	der                     *pem.Block
+	stdDialer               *cluster.StandardDialer
+	clusterVerifyBlocks     ClusterVerifyBlocksFunc // Default: cluster.VerifyBlocks, or a mock for testing
+}
+
+// ClusterVerifyBlocksFunc is a function that matches the signature of cluster.VerifyBlocks, and allows mocks for testing.
+type ClusterVerifyBlocksFunc func(blockBuff []*common.Block, signatureVerifier cluster.BlockVerifier) error
+
+// NewBlockPullerFactory creates a new BlockPullerFactory, using the configuration details that do not change during
+// the life cycle of the orderer.
+func NewBlockPullerFactory(
 	channelID string,
+	logger *flogging.FabricLogger,
 	signer identity.SignerSerializer,
-	blockVerifier cluster.BlockVerifier,
 	baseDialer *cluster.PredicateDialer,
 	clusterConfig localconfig.Cluster,
 	bccsp bccsp.BCCSP,
-) (*cluster.BlockPuller, error) {
+	alternativeVerifyBlocksFunc ...ClusterVerifyBlocksFunc,
```

Comment:
This strikes me as really odd.

I understand that by making this a varargs, you can omit it and things work 'with defaults', but it's a hack, and misleading, as you can't meaningfully specify more than one.  I've seen this done as a hack in places where we really don't want to break the API because it would require a cascade of changes or it's a public API, but in this case, we're already breaking it by adding the logger, so I don't understand the value.

It would actually seem much more natural to me to do away with the constructor, and instead have a struct with public members that are set, and give you the 'expected' defaults when they are omitted.  This has the bonus of making the construction easier to read as it becomes clear what each argument is meant to do.  Barring a more radical change like the above, I would certainly get rid of the varargs.

Reply:
I got rid of the varargs.

I would like to keep the constructor because there is a bit of processing going on in it when building a `BlockPullerFactory` that I would like to hide from the follower creation method in the multichannel package.